### PR TITLE
Custom API for Acme Inc 

### PR DESCRIPTION
**Pull request by ccstevens**
Adds a SQL Statement Execution API demo that builds a custom API using a Python Flask backend and an HTML/JQuery frontend.

### CODE REVIEW

File: custom-api/store_server.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+     SqlStatement.LIST_STORES: """
+            select
+                s_store_sk as store_id, 
+                s_store_name as store_name, 
+                s_manager as manager,
+                s_number_employees as number_employees, 
+                s_city as city,
+                s_state as state
+            from 
+                chris_stevens.retail.store
```
Comment (ccstevens) [inline custom-api/store_server.py:39]: Obviously, this needs to change.

File: custom-api/templates/store.html [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+      // Initial load of the last 20 sales for the store in JSON format.
+      showLoadingIcon();
+      getSalesAsync("JSON_ARRAY", 20, function(requestId) {
+        if (requestId !== null) {
+          setTimeout(
+            function() { 
+              pollSalesStatus(requestId, displaySales) 
+            }, 
+            100 // 100ms polling interval.
```
Comment (kjhuang-db) [inline custom-api/templates/store.html:118]: nit: for demo purpose I think it is fine`setTimeout` second argument is not "polling" interval though (`setInterval` is the polling)

File: custom-api/templates/store.html [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+            results.push(result);
+            return next();
+          });
+        }
+
+        const limitedPromises = Array(concurrencyLimit)
+          .fill(null)
+          .map(() => next());
+
+        return Promise.all(limitedPromises).then(() => results);
```
Comment (kjhuang-db) [inline custom-api/templates/store.html:248]: does this mean we will not show all result chunks?

File: custom-api/templates/store.html [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+      // Get an external link and download the data.
+      async function downloadFromExternalLink(requestId, chunkIndex) {
+        const externalLink = await getExternalLink(requestId, chunkIndex);
+        const data = await $.ajax({
+          url: externalLink,
+          type: "GET"
+        });
+        return new Promise((resolve, reject) => {
+          resolve({chunkIndex, data});
+        });
```
Comment (kjhuang-db) [inline custom-api/templates/store.html:103]: nit: since we already decorate it as `async` we don't need to return a self resolve promise, we can just return `{chunkIndex, data}`.

**Review by kjhuang-db — APPROVED** [review]
pull down locally and it works! 

for js code, the mix usage of "async await", promise or callback styles could be more unified, but for demo purpose I think it is fine

File: custom-api/templates/store.html (hunk 3 above)
Comment (ccstevens) [reply custom-api/templates/store.html:248]: This does end up getting the full set of data. I get 500MB CSV files and expect that large. There is some magic here with the promises where one completed triggers another one from the queue.

File: custom-api/templates/store.html (hunk 4 above)
Comment (ccstevens) [reply custom-api/templates/store.html:103]: It didn't work for me. Let me try again.

File: custom-api/store_server.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+    statement_response = w.statement_execution.execute_statement(
+        statement = sql_statements[SqlStatement.LIST_SALES],
+        format = Format[format],
+        disposition = Disposition.EXTERNAL_LINKS,
+        wait_timeout = "0s",
+        warehouse_id = warehouse_id,
+        parameters = parameters,
+        row_limit = row_limit,
+        byte_limit = 500000000 # 500MB limit because lcp crashes.
```
Comment (sander-goos) [inline custom-api/store_server.py:184]: nit: lcp -> local cors proxy?

File: custom-api/store_server.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+    response = {
+        'request_id': statement_response.statement_id,
+        'state': str(statement_response.status.state.name)
+    }
+
+    return response
+
+# Poll the Statement Execution API for status on a previously generated execution request.
+def get_list_sales_status(request_id):
```
Comment (sander-goos) [inline custom-api/store_server.py:195]: nit: this method has nothing specific related to list sales so we can use a more generic name

File: custom-api/store_server.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        'state': str(statement_response.status.state.name)
+    }
+
+    if statement_response.status.state == StatementState.SUCCEEDED:
+        response['chunk_count'] = statement_response.manifest.total_chunk_count
+
+    return response
+
+# Get the external link for the given chunk. The pre-signed URL is valid for 15 mins.
+def get_list_sales_chunk(request_id, chunk_index):
```
Comment (sander-goos) [inline custom-api/store_server.py:208]: nit: same here, we could use a more generic name

File: custom-api/store_server.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+#
+# WARNING: Don't use in production.
+# 
+# The following is needed to overcome CORS without Databricks workspace bucket modifications.
+#
+local_cors_proxy_lock = threading.Lock()
+local_cors_proxy_enabled = False
+
+def get_host(external_link):
```
Comment (sander-goos) [inline custom-api/store_server.py:272]: Can we use https://docs.python.org/3/library/urllib.parse.html instead?

File: custom-api/store_server.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        StatementParameterListItem(name='store_id', value=store_id, type="INT"),
+        StatementParameterListItem(name='item_id', value=item_id, type="INT"),
+        StatementParameterListItem(name='quantity', value=quantity, type="INT"),
+        StatementParameterListItem(name='price', value=sales_price, type="DECIMAL(7,2)")
+    ]
+
+    statement_response = w.statement_execution.execute_statement(
+        statement = sql_statements[SqlStatement.INSERT_SALE],
+        wait_timeout = "50s",
+        on_wait_timeout = TimeoutAction.CONTINUE,
```
Comment (sander-goos) [inline custom-api/store_server.py:252]: QQ: I don't think we poll if this doesn't complete within 50s right?

File: custom-api/create_data.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  if response.status.state != StatementState.SUCCEEDED:
+    print("Failed to execute: '" + sql_statement + "' with state " + str(response.status.state));
+    if response.status.state == StatementState.FAILED:
+        print("Detailed error message: " + response.status.error.message)
+    exit()
+
+  return
+
+# Create a database.
+print("Creating acme_demo database.")
```
Comment (sander-goos) [inline custom-api/create_data.py:34]: Should we check if it already exists and bail out if so?
Currently it fails with an error and you are not sure if the rest of the setup is done.

File: custom-api/templates/new_sale.html [hunk 11]
```diff
@@ -0,0 +1,59 @@
+<!DOCTYPE html>
```
Comment (sander-goos) [inline custom-api/templates/new_sale.html:1]: How do we get to this page? I was expecting a link from the store page?

File: custom-api/README.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+1. Run the setup script: `source ./setup.sh`
+    - Installs Flask
+    - Installs the Databricks SDK for Python
+    - Installs [Local CORS Proxy](https://www.npmjs.com/package/local-cors-proxy)
+        - This is use to overcome [CORS](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) errors in the demo. Not required for production.
+    - Initializes Databricks host, token, and Warehouse ID environment variables.
+    - Creates sample tables in your hive metastore. The size is < 50KB.
+        - This is a significantly smaller data set that we used in the Data+AI Summit demo.
+2. Run the app: `python3 store_server.py`
+3. Visit http://127.0.0.1:5000 in your browser.
```
Comment (sander-goos) [inline custom-api/README.md:29]: Works on my machine!

File: custom-api/store_server.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    return response
+
+# Poll the Statement Execution API for status on a previously generated execution request.
+def get_list_sales_status(request_id):
+    statement_response = w.statement_execution.get_statement(statement_id = request_id)
+    response = {
+        'request_id': statement_response.statement_id,
+        'state': str(statement_response.status.state.name)
```
Comment (sander-goos) [inline custom-api/store_server.py:199]: Because we're not returning the result data here (which is available) it costs another round trip to fetch the first chunk.

File: custom-api/store_server.py [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+    # Use parameters to prevent SQL injection via the store ID string.
+    parameters = [
+        StatementParameterListItem(name='store_id', value=store_id, type="INT")
+    ]
+
+    statement_response = w.statement_execution.execute_statement(
+        statement = sql_statements[SqlStatement.LIST_SALES],
+        format = Format[format],
+        disposition = Disposition.EXTERNAL_LINKS,
+        wait_timeout = "0s",
```
[the remark this answers is not in the document]
Comment (ccstevens) [reply custom-api/store_server.py:180]: Yeah. It does seem about double the page load time, even for me. ~800ms for 100 stores and ~1200-1600ms for 20 sales. And it is the 2 extra round trips. We could optimize away one of them.

I was trying to not have the disposition as part of the API, but could add that. It would be less code sharing on the FE though.

File: custom-api/store_server.py (hunk 7 above)
Comment (ccstevens) [reply custom-api/store_server.py:208]: Will fix.

File: custom-api/store_server.py (hunk 9 above)
Comment (ccstevens) [reply custom-api/store_server.py:252]: Nope. It's fire and forget 😄 . Lazy on the FE. Should probably poll and see it to a terminal state.

File: custom-api/templates/store.html (hunk 2 above)
Comment (ccstevens) [reply custom-api/templates/store.html:118]: oh, I get what you're saying. It's the delay before executing the function. So, it effectively creates a polling interval, but not 100% accurate to say what I wrote.

File: custom-api/templates/store.html (hunk 4 above)
Comment (ccstevens) [reply custom-api/templates/store.html:103]: ok. works now 🤷 thanks for catching!

File: custom-api/create_data.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+execute_statement_sync("create database hive_metastore.acme_demo")
```
Comment (svar29) [inline custom-api/create_data.py:35]: I get this when running setup
```
Failed to execute: 'create database hive_metastore.acme_demo' with state StatementState.FAILED
Detailed error message: [INSUFFICIENT_PERMISSIONS] Insufficient privileges:
User does not have permission CREATE on CATALOG.
```
Am I missing some permissions in dogfood?
I guess I would need create catalog permissions
https://docs.databricks.com/en/data-governance/unity-catalog/manage-privileges/privileges.html

I was able to set this up on BUI dev workspace where I am an admin.

File: custom-api/acme_server.py [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+# It takes no arguments and invokes the Statement Execution API in synchronous mode, returning store data results in
+# JSON format. If the request to Databricks SQL takes longer than 50 seconds, the execution will be cancelled and no
+# data is returned.
+#
+@app.route("/api/1.0/stores", methods=["GET"])
+def list_stores():
+    statement_response = w.statement_execution.execute_statement(
+        statement = sql_statements[SqlStatement.LIST_STORES],
+        wait_timeout = "50s",
+        on_wait_timeout = TimeoutAction.CANCEL,
```
Comment (svar29) [inline custom-api/acme_server.py:114]: Got this error. 
```
[2023-09-20 22:19:18,347] ERROR in app: Exception on /api/1.0/stores [GET]
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/varun.sharma/databricks/dbsql-rest-api/custom-api/acme_server.py", line 114, in list_stores
    on_wait_timeout = TimeoutAction.CANCEL,
                      ^^^^^^^^^^^^^
NameError: name 'TimeoutAction' is not defined
```

I think it should be ExecuteStatementRequestOnWaitTimeout instead as changing code to that helped.

File: custom-api/templates/new_sale.html [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+  <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/styles.css') }}">
+  <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
+  <script>
+    $(document).ready(function() {
+      $('#sale-form').submit(function(event) {
+        event.preventDefault(); // Prevent the default form submission
+
+        var form = document.getElementById("sale-form");
+        console.log("storeId: " + form.elements.storeId.value);
+        var formData = new FormData(form);
```
Comment (svar29) [inline custom-api/templates/new_sale.html:14]: when I entered the date wrongly formatted, it failed to insert without any errors in UI or in flask server.

**Review by svar29 — COMMENTED** [review]
Thanks for prepping this demo. I tried it out and left some comments on problems I faced.

File: custom-api/acme_server.py (hunk 16 above)
Comment (ccstevens) [reply custom-api/acme_server.py:114]: Seems like a regression in the SDK.

File: custom-api/create_data.py (hunk 15 above)
Comment (ccstevens) [reply custom-api/create_data.py:35]: Hmm. Only admins can create in the hive_metastore?

File: custom-api/setup.sh [hunk 18]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+# Install dependencies.
+pip3 install flask
+pip3 install databricks-sdk
+npm install -g local-cors-proxy
+
+# Get the Databricks HOST.
+echo "Enter your Databricks HOST (i.e. foo.databricks.com):"
```
[the remark this answers is not in the document]
Comment (ccstevens) [reply custom-api/setup.sh:9]: hmm. My setup I only have `DATABRICKS_HOST=e2-dogfood.staging.cloud.databricks.com`

Could be another SDK change.

File: custom-api/templates/new_sale.html (hunk 17 above)
Comment (ccstevens) [reply custom-api/templates/new_sale.html:14]: Define wrongly formated?

File: custom-api/setup.sh [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+read HOST
+echo
+export DATABRICKS_HOST=$HOST
+
+# Get the access token.
+echo "Enter your Databricks personal access token:"
+read TOKEN
```
[the remark this answers is not in the document]
Comment (ccstevens) [reply custom-api/setup.sh:16]: `read -s` does the trick without needing python

File: custom-api/setup.sh (hunk 18 above)
[the remark this answers is not in the document]
Comment (ccstevens) [reply custom-api/setup.sh:9]: Looks like the SDK now handles both: https://github.com/databricks/databricks-sdk-py/blob/349216706aeac81828a807f40a21a1b0c80ed717/databricks/sdk/core.py#L783

[the remark this answers is not in the document]
Comment (ccstevens) [reply custom-api/setup.sh:9]: Hmm. That's been there for 6 months.

[the remark this answers is not in the document]
Comment (ccstevens) [reply custom-api/setup.sh:9]: ah... you had a typo: `'e2-dogfood.staging.cloud.databricks.comm`. 2 m's at the end.

File: custom-api/templates/new_sale.html (hunk 17 above)
Comment (ccstevens) [reply custom-api/templates/new_sale.html:14]: never mind. I get it. Can fix.
